src/OpenTelemetry/Metrics/MetricPoint.cs

```diff
@@ -337,16 +337,18 @@ internal void Update(double number)
                         var sw = default(SpinWait);
                         while (true)
                         {
-                            if (Interlocked.Exchange(ref this.histogramBuckets.IsCriticalSectionOccupied, 1) == 0)
+                            if (Interlocked.CompareExchange(ref this.histogramBuckets.IsCriticalSectionOccupied, 1, 0) == 0)
```

Comment:
Just curious if there were more learnings based on the questions raised here https://github.com/open-telemetry/opentelemetry-dotnet/pull/3458#discussion_r925842839.

Does `CompareExchange` help alleviate the concerns? Or did you learn the concern was unfounded?

Reply:
I introduced `CompareExchange` in this PR because I think it's much better to try to update `IsCriticalSectionOccupied` to 1 only when it's actually 0. With the `Interlocked.Exchange` we would be setting that variable to 1 even when it's already set to 1, and that seems wasteful.

Regarding the learnings for the questions asked in #3458:

1. We can use Interlocked operations to devise a "lock" that would synchronize the access to histogram related fields here among multiple threads. However, the part where we release this "lock" i.e., where we set `IsCriticalSectionOccupied` = 0 has to be done using an `Interlocked` statement as well. This is because `Interlocked` creates a memory fence around that variable. This prevents certain reorderings of instructions that might set `IsCriticalSection` to 0 before the original thread is done processing the histogram fields as that might allow another thread to enter the critical section while the original thread has not exited it.

2. Also, any other thread which might be accessing these histogram fields must use the same synchronization mechanism. That's why I have updated the `TakeSnapshot` method to use the same Interlocked operations to update `IsCriticalSectionOccupied` to determine if it's safe to read these fields.

3. The instructions within this "lock" can be reordered, in this case, the code under `unchecked`. That's okay because we don't care what order they get executed in.

4. Regarding the freshness of the values, as long as all the threads that access the same set of fields use the same synchronization mechanism (could be the C# `lock` statement or our Interlocked based "lock"), any thread that enters the critical section would see the most up-to-date value of these fields.

5. The only thing to be wary of in this Interlocked based "lock" is exceptions. If an exception is thrown while a thread was in the critical section, `IsCriticalSectionOccupied` would not be set back to 0 and other threads waiting to enter the critical section would never be able to enter it. In our case, the instructions inside the "lock" would not throw an exception so we don't have to worry about that.

Reply:
> I introduced CompareExchange in this PR because I think it's much better to try to update IsCriticalSectionOccupied to 1 only when it's actually 0. With the Interlocked.Exchange we would be setting that variable to 1 even when it's already set to 1, and that seems wasteful.

^ might be good to see if perf is impacted with this.

Reply:
Excellent summary! Thank you. Number 5 in particular had not occurred to me, but absolutely makes sense.

I'd be interested in perf too. Given that the snapshot and update threads were previously not synchronizing with one another, I'm curious if perf is impacted now that they are correctly synchronized. Maybe not by much, since the snapshot thread only runs seldomly...

Reply:
I'm OK with this as-is but we should probably do some perf testing using `lock`, `SpinLock`, and this new `Interlocked` code. I think this will be the fastest way but I don't want to blindly trust the old test which proved "Interlocked is faster than lock" because it was bugged when tested 😄

PS: Special shout-out to @noahfalk for helping us on this 👏

Reply:
for correctness, we need to modify this tests https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs#L1175 
It validates multiple threads writing scenario.
We can add a reader thread also, and ultimately everything should add up correctly. (ideally we want to modify the stress tests to do it, so you can run it for hours and at the end, it'll confirm if the metrics are correct).